src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/LoggingObserver.cs

```diff
@@ -0,0 +1,102 @@
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using RegressionGames.StateRecorder.Models;
+using UnityEngine;
+
+namespace RegressionGames.StateRecorder
+{
+    /// <summary>
+    /// Watches for logging messages during recording
+    /// </summary>
+    public class LoggingObserver : MonoBehaviour
+    {
+        public long LoggedWarnings { get; private set;  }
+        public long LoggedErrors { get; private set; }
+        private ConcurrentQueue<LogDataPoint> _logQueue = new();
+        private Task _fileWriteTask;
+
+        public void StartCapturingLogs()
+        {
+            LoggedWarnings = 0;
+            LoggedErrors = 0;
+            _fileWriteTask = null;
+            Application.logMessageReceivedThreaded += OnLogMessageReceived;
+        }
+        
+        public void StopCapturingLogs()
+        {
+            Application.logMessageReceivedThreaded -= OnLogMessageReceived;
+            
+            // wait for any pending write tasks to complete
+            if (_fileWriteTask != null)
+            {
+                _fileWriteTask.Wait();
+            }
+        }
+        
+        private void OnLogMessageReceived(string message, string stacktrace, LogType type)
+        {
+            if (type == LogType.Warning)
+            {
+                LoggedWarnings++;
+            } else if (type == LogType.Error || type == LogType.Exception)
+            {
+                LoggedErrors++;
+            }
+            
+            // PERF: DateTimeOffset.Now is surprisingly costly, use DateTimeOffset.UtcNow unless we have to know the local timezone.
+            // It's still good to use a DateTimeOffset here though because it ensures that when we serialize the data,
+            // it is clearly marked as a UTC time (with the `Z` suffix)
+            var dataPoint = new LogDataPoint(
+                timestamp: DateTimeOffset.UtcNow,
+                frameTimestamp: Time.unscaledTimeAsDouble,
+                level: type, 
+                message: message, 
+                stackTrace: stacktrace
+            );
+            _logQueue.Enqueue(dataPoint);
+        }
+        
+        /// <summary>
+        /// Called each tick to flush any available logs to disk.
+        /// We may not actually write logs to file every tick:
+        /// If there's a running task accessing the destination file, then we'll skip this tick and try again next tick.
+        /// </summary>
+        public void FlushLogs(string filePath, CancellationToken token)
+        {
+            if (_fileWriteTask != null && !_fileWriteTask.IsCompleted)
```

Comment:
The internals of this method need to be wrapped in a `lock` statement.

If this method were to be called 2x close in time, it can start 2 file write tasks to the same file in parallel.  The reason for this is that both calls can get past this first check thinking the current fileWriteTask is completed because there is no locking to ensure one of them created a new task before the other one checks.